File: src/main/java/nl/anchormen/sbt/AbiBin.java
package nl.anchormen.sbt;

import java.io.File;
import java.io.IOException;
import java.io.UncheckedIOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class AbiBin
{

    private final Path searchBase;

    public final String name;
    public final Path basePath;


    private Path abi;
    private Path bin;
    private Path relativePath;

    public AbiBin(Path searchBase, Path abiBinPath)
    {
        this.searchBase = searchBase;
        this.basePath = abiBinPath.getParent();
        this.name = removeExtension(abiBinPath.getFileName().toString());
    }

    public Path abi()
    {
        if (this.abi == null)
        {
            this.abi = this.basePath.resolve(this.name + ".abi");
        }
        return this.abi;
    }

    public Path bin()
    {
        if (this.bin == null)
        {
            this.bin = this.basePath.resolve(this.name + ".bin");
        }
        return this.bin;
    }

    public Path relativePath()
    {
        if (this.relativePath == null)
        {
            this.relativePath = this.searchBase.relativize(this.basePath);
        }
        return this.relativePath;
    }

    public String packageName()
    {
        Path p = relativePath();
        int l = p.getNameCount();
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < l; i++)
        {
            sb.append(p.getName(i));
            if (i < l - 1)
            {
                sb.append('.');
            }
        }
        return sb.toString();
    }


    public boolean isValid()
    {
        return Files.isRegularFile(abi()) && Files.isRegularFile(bin());
    }

    @Override
    public boolean equals(Object o)
    {
        if (this == o)
        {
            return true;
        }
        if (o == null || getClass() != o.getClass())
        {
            return false;
        }
        AbiBin abiBin = (AbiBin) o;
        return Objects.equals(name, abiBin.name) &&
                Objects.equals(basePath, abiBin.basePath);
    }

    @Override
    public int hashCode()
    {
        return Objects.hash(name, basePath);
    }

    public static Stream<AbiBin> find(Path basePath)
    {
        Path ensureFullPath = basePath.toAbsolutePath();
        try
        {
            return Files.walk(ensureFullPath)
                    .filter(p -> p.getFileName().toString().endsWith(".abi") || p.getFileName().toString().endsWith(".bin"))
                    .map(p -> new AbiBin(ensureFullPath, p))
                    .distinct()
                    .filter(AbiBin::isValid);
        }
        catch (IOException ex)
        {
            throw new UncheckedIOException(ex);
        }
    }

    public static List<AbiBin> findList(File f)
    {
        return findList(f.toPath());
    }

    public static List<AbiBin> findList(Path path)
    {
        return find(path)
                .collect(Collectors.toList());
    }

    private static String removeExtension(String name)
    {
        return name.substring(0, name.length() - 4);
    }
}

File: src/main/scala/nl/anchormen/sbt/AbiBin.scala
package nl.anchormen.sbt

import java.io.{File, IOException, UncheckedIOException}
import java.nio.file.{Files, Path, Paths}
import java.util.regex.Pattern

object AbiBin {
	val AbiExtension: String = ".abi"
	val BinExtension: String = ".bin"
	val ExtensionLength: Int = 4

	def findList(f: File): List[AbiBin] = {
		findListImpl(f.toPath)
	}

	def findList(path: Path): List[AbiBin] = {
		findListImpl(path)
	}

	def convertPathToPackageWithFileName(path: Path): String = {
		val nameCount: Int = path.getNameCount

		val pathNames: Seq[Path] = for (index <- 0 until nameCount) yield {
			val name: Path = path.getName(index)

			if (index == nameCount - 1) {
				Paths.get(removeExtension(name.toString))
			} else {
				name
			}
		}

		pathNames.mkString(".")
	}

	def getFileNameFromPackageName(packageName: String): String = {
		packageName.split(Pattern.quote(".")).last
	}

	private def hasCorrectExtension(path: Path): Boolean = {
		path.getFileName.toString.endsWith(AbiExtension) || path.getFileName.toString.endsWith(BinExtension)
	}

	private def findListImpl(basePath: Path): List[AbiBin] = {
		val baseSearchPath = basePath.toAbsolutePath

		try {
			val pf: PartialFunction[Path, (String, Path)] = {
				case p: Path => {
					val relativePath: Path = baseSearchPath.getParent.relativize(p)
					(convertPathToPackageWithFileName(relativePath), p)
				}
			}

			import scala.collection.JavaConverters._

			Files.walk(baseSearchPath)
					.iterator()
					.asScala
					.filter(hasCorrectExtension)
					//Collect the file paths based on their package name, creating a tuple (name, path)
					.collect(pf)
					.toSeq
					//Group by package name
					.groupBy(_._1)
					.mapValues(_.map(_._2))
					//Filter on map entries that have 2 values, since we want both a .abi and a .bin file
					.filter(k => k._2.lengthCompare(2) == 0)
					.map {
						entry: (String, Seq[Path]) => new AbiBin(entry._1, entry._2.head, baseSearchPath)
					}.toList
		} catch {
			case ex: IOException => throw new UncheckedIOException(ex)
		}
	}

	private def removeExtension(name: String) = {
		name.substring(0, name.length - ExtensionLength)
	}
}

class AbiBin(val packageWithFileName: String, val abiBinPath: Path, val searchBase: Path) {
	lazy private val _name: String = AbiBin.getFileNameFromPackageName(packageWithFileName)
	lazy private val _packageName: String = getPackageNameFromPackageWithFileName(packageWithFileName)
	lazy private val _basePath: Path = abiBinPath.getParent
	lazy private val _abi: Path = _basePath.resolve(_name + AbiBin.AbiExtension)
	lazy private val _bin: Path = _basePath.resolve(_name + AbiBin.BinExtension)

	def name: String = _name
	def packageName: String = _packageName
	def abi: Path = _abi
	def basePath: Path = _basePath
	def bin: Path = _bin

	def isValid: Boolean = Files.isRegularFile(_abi) && Files.isRegularFile(_bin)
	def newLocation(out: File): File = out.toPath.resolve(relativePath).resolve(_name + ".java").toFile

	private lazy val relativePath: Path = {
		val path = searchBase.relativize(_basePath)
		val initialPathName = searchBase.getName(searchBase.getNameCount - 1)

		if (Files.isSameFile(searchBase.toAbsolutePath, _basePath.toAbsolutePath)) {
			initialPathName
		} else {
			initialPathName.resolve(path)
		}
	}

	private def getPackageNameFromPackageWithFileName(packageWithFileName: String): String = {
		packageWithFileName.split(Pattern.quote(".")).dropRight(1).mkString(".")
	}
}
